src/cmd/script.rs

```diff
@@ -781,542 +483,643 @@ fn emit_synopsys_tcl(
-                    }
-                    println!();
-                    println!("{}", lines.join(" \\\n    "));
-                    if abort_on_error {
-                        println!("{}", tcl_catch_postfix());
-                    }
-                },
-            );
-        }
-    } else {
-        let mut lines = vec![];
-        let mut file_lines = vec![];
-        let mut inc_dirs = Vec::new();
-        let mut files = vec![];
-        let mut defines: Vec<(String, Option<String>)> = vec![];
-        let mut t: bool = false;
-        for src in srcs {
-            inc_dirs = src.clone().get_incdirs();
-            files.append(&mut src.files.clone());
-            defines.extend(
-                src.defines
-                    .iter()
-                    .map(|(k, &v)| (k.to_string(), v.map(String::from))),
-            );
-        }
-        for file in files {
-            let p = match file {
-                SourceFile::File(p) => p,
-                _ => continue,
-            };
-            if p.to_str().unwrap().contains(".sv")
-                || p.to_str().unwrap().contains(".v")
-                || p.to_str().unwrap().contains(".vp")
-            {
-                t = true;
-            } else if p.to_str().unwrap().contains(".vhd") || p.to_str().unwrap().contains(".vhdl")
-            {
-                t = false;
-            }
-            file_lines.push(quote(&relativize_path(p, sess.root)));
-        }
-        if t {
-            lines.push(tcl_catch_prefix("vlog -sv", abort_on_error));
-            if let Some(args) = matches.get_many::<String>("vlog-arg") {
-                lines.extend(args.map(Into::into));
-            }
-            defines.extend(
-                targets
-                    .iter()
-                    .map(|t| (format!("TARGET_{}", t.to_uppercase()), None)),
-            );
-            add_defines_from_matches(&mut defines, matches);
-            defines.sort();
-            for (k, v) in defines {
-                let mut s = format!("+define+{}", k.to_uppercase());
-                if let Some(v) = v {
-                    s.push('=');
-                    s.push_str(&v);
-                }
-                lines.push(s);
-            }
+static FLIST_TPL: &str = "\
+{% for incdir in all_incdirs %}\
+    {% if relativize_path %}\
+        +incdir+{{ incdir | replace(from=root, to='') }}\n\
+    {% else %}\
+        +incdir+{{ incdir }}\n\
+    {% endif %}\
+{% endfor %}\
+{% for define in all_defines %}\
+    +define+{{ define.0 | upper }}{% if define.1 %}={{ define.1 }}{% endif %}\n\
+{% endfor %}\
+{% for file in all_files %}\
+    {% if relativize_path %}\
+        {% if file is starting_with(root) %}\
+            {{ file | replace(from=root, to='') | trim_start_matches(pat='/') }}\n\
+        {% else %}\
+            {{ file }}\n\
+        {% endif %}\
+    {% else %}\
+        {{ file }}\n\
+    {% endif %}\
+{% endfor %}";
```

Comment:
May this not be easier with `include_str`?

Reply:
I originally did this with a string internal to the Rust file for clearer formatting of the template, but found a way for semi-clean formatting with comments also outside of the file, included now with `include_str`. This can also provide a better basis for custom templates, in case they are needed.